labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java

```diff
@@ -136,6 +137,66 @@ public void createLabel(
     }
   }
 
+  @Override
+  public void getOrCreateLabels(
+      GetOrCreateLabelsRequest request,
+      StreamObserver<GetOrCreateLabelsResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      Lock labelsLock = this.stripedLabelsLock.get(requestContext.getTenantId());
+      if (labelsLock.tryLock(WAIT_TIME.getSeconds(), TimeUnit.SECONDS)) {
+        try {
+          final Map<String, Label> existingLabelsMap = getLabelsMap();
+          final List<Label> newLabels = new ArrayList<>();
+          final List<Label> allLabels = new ArrayList<>();
+          request.getRequestsList().stream()
+              .forEach(
+                  labelRequest ->
+                      this.buildLabelFromRequest(
+                          labelRequest, existingLabelsMap, newLabels, allLabels));
+          if (!newLabels.isEmpty()) {
+            List<Label> createdLabels =
+                labelStore.upsertObjects(requestContext, newLabels).stream()
+                    .map(org.hypertrace.config.objectstore.ConfigObject::getData)
+                    .collect(Collectors.toList());
+            allLabels.addAll(createdLabels);
+          }
+          responseObserver.onNext(
+              GetOrCreateLabelsResponse.newBuilder().addAllLabels(allLabels).build());
```

Comment:
nit - can we preserve the order of labels on the response to match that of the request? That'd be my assumption as a caller

Reply:
okay. Made this change.